caravan.lua

```diff
@@ -1,31 +1,311 @@
+                end
+
+                if collapsed_this_container then
+                    num_items_collapsed = num_items_collapsed + #dfhack.items.getContainedItems(df.global.game.main_interface.trade.good[list_index][k])
+                end
+            end
+
+            ::skip::
+        end
+    end
+
+    if num_items_collapsed > 0 then
+        df.global.game.main_interface.trade.i_height[list_index] = df.global.game.main_interface.trade.i_height[list_index] - num_items_collapsed * 3
+    end
+end
+
+function collapseAllContainers()
+    collapseContainers(df.global.game.main_interface.trade.goodflag[0], 0)
+    collapseContainers(df.global.game.main_interface.trade.goodflag[1], 1)
+end
+
+function collapseEverything()
+    collapseAllContainers()
+    collapseAllTypes()
+end
+
+function copyGoodflagState()
+    trader_selected_state = copyall(df.global.game.main_interface.trade.goodflag[0])
+    broker_selected_state = copyall(df.global.game.main_interface.trade.goodflag[1])
+end
+
+function CaravanOverlay:onInput(keys)
+    if not is_trade_overlay_active() then return false end
+    if keys._MOUSE_L then
+        if dfhack.internal.getModifiers().shift then
+            handle_shift_click_on_render = true
+        elseif dfhack.internal.getModifiers().ctrl then
+            handle_ctrl_click_on_render = true
+        end
+
+        if handle_ctrl_click_on_render or handle_shift_click_on_render then
+            copyGoodflagState()
+        end
+    end
+    CaravanOverlay.super.onInput(self, keys)
+end
+
+function CaravanOverlay:render(dc)
+    if not is_trade_overlay_active() then return false end
+    if handle_shift_click_on_render then
+        handle_shift_click_on_render = false
+        select_shift_clicked_container_items(df.global.game.main_interface.trade.goodflag[0], trader_selected_state, 0)
+        select_shift_clicked_container_items(df.global.game.main_interface.trade.goodflag[1], broker_selected_state, 1)
+    elseif handle_ctrl_click_on_render then
+        handle_ctrl_click_on_render = false
+        collapse_ctrl_clicked_containers(df.global.game.main_interface.trade.goodflag[0], trader_selected_state, 0)
+        collapse_ctrl_clicked_containers(df.global.game.main_interface.trade.goodflag[1], broker_selected_state, 1)
+    end
+
+    CaravanOverlay.super.render(self, dc)
+end
+
+
+CaravanInfoOverlay = defclass(CaravanInfoOverlay, overlay.OverlayWidget)
+CaravanInfoOverlay.ATTRS{
+    default_pos={x=-3,y=-12},
+    default_enabled=true,
+    viewscreens='dwarfmode',
+    frame={w=27, h=13},
+    frame_style=gui.MEDIUM_FRAME,
+    frame_background=gui.CLEAR_PEN,
+}
+
+function CaravanInfoOverlay:init()
+    self:addviews{
+        widgets.Label{
+            frame={t=0, l=0},
+            text = 'Shift+Click checkbox:',
+            text_pen=COLOR_LIGHTGREEN,
+        },
+        widgets.Label{
+            frame={t=1, l=0},
+            text = 'select items inside bin',
+        },
```

Comment:
not like what you have is wrong, but just so you know, Labels are quite flexible with their text. This could also be written like this:
```lua
widgets.Label{
    frame={t=0, l=0},
    text={
        {text='Shift+Click checkbox:', pen=COLOR_LIGHTGREEN},
        NEWLINE,
        'select items inside bin',
    },